File: src/main/java/core/procedures/strings/Replace.java
package core.procedures.strings;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.regex.Pattern;

public final class Replace extends AFn {

  public Replace() {
    super(new FnArgsBuilder().min(3).max(3).build());
  }

  @Override
  public String getName() {
    return "replace";
  }

  @Override
  public String apply3(Object arg1, Object arg2, Object arg3) {
    if (!(arg1 instanceof CharSequence)) {
      throw new WrongTypeException(getName(), "String", arg1);
    }
    if (arg2 instanceof Character && arg3 instanceof Character) {
      return arg1.toString().replace((Character) arg2, (Character)arg3);
    }
    if (arg2 instanceof CharSequence && arg3 instanceof CharSequence) {
      return arg1.toString().replace((CharSequence) arg2, (CharSequence)arg3);
    }
    // TODO arg2=string/arg3=function of match
    return ((Pattern)arg2).matcher((CharSequence)arg1).replaceAll(((CharSequence)arg3).toString());
  }
}

File: src/main/java/core/procedures/strings/Replace.kt
package core.procedures.strings

import core.exceptions.WrongTypeException
import core.procedures.AFn

import java.util.regex.Pattern

class Replace : AFn(name = "replace", isPure = true, minArgs = 3, maxArgs = 3) {

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?): String? {
        if (arg1 !is CharSequence) {
            throw WrongTypeException(name, "String", arg1)
        }
        if (arg2 is Char && arg3 is Char) {
            return arg1.toString().replace((arg2 as Char?)!!, (arg3 as Char?)!!)
        }
        if (arg2 is CharSequence && arg3 is CharSequence) {
            return arg1.toString().replace(arg2.toString(), arg3.toString())
        }
        // TODO arg2=string/arg3=function of match
        return (arg2 as Pattern).matcher((arg1 as CharSequence?)!!).replaceAll((arg3 as CharSequence).toString())
    }
}
